astropy/cosmology/io.py

```diff
@@ -0,0 +1,336 @@
+            close_after = False
+        elif isinstance(file, (dict, Mapping)):
+            return file
+        else:
+            raise TypeError("`file` must be path-like or file-like.")
+
+        # ---- Read -----
+        params = json.loads(file.read())
+        if close_after:
+            file.close()
+        return params
+
+    def from_mapping(self, mapping, key=None):
+        """Load `~astropy.cosmology.Cosmology` from mapping object.
+
+        Parameters
+        ----------
+        mapping : mapping
+        key : hashable or None, optional
+            Any valid key into `mapping`.
+            If not None, this is used to select a cosmology if mapping is a
+            set of cosmologies.
+
+        Returns
+        -------
+        `~astropy.cosmology.Cosmology`
+
+        """
+        # get params subset from mpping
+        params = mapping[key] if key is not None else mapping
+
+        # determine class
+        subclasses = dict(self._cosmo_cls.__all_subclasses__())
+        cls_name = params.pop("cosmology", self._cosmo_cls.__qualname__)
+        try:
+            cosmo_cls = subclasses[cls_name]
+        except KeyError:
+            raise KeyError(
+                f"{cls_name} is not a subclass of {cls.__qualname__}.")
+
+        # match keys to arguments
+        inspect_params = cosmo_cls._init_signature.parameters.values()
+        expected_params = set([
+            p.name for p in inspect_params
+            if not ((p.kind == _VAR_POSITIONAL) or (p.kind == _VAR_KEYWORD))])
+        HAS_KWARGS = any([p.kind == _VAR_KEYWORD for p in inspect_params])
+
+        # move non-parameters to metadata, if no **kwargs in signature
+        meta = params.pop("meta", None) or {}
+        for k, v in tuple(params.items()):
+            if k not in expected_params and not HAS_KWARGS:
+                v = params.pop(k)  # make sure not in params
+                # TODO! raise an error / warning if repeat k?
+                meta.setdefault(k, v)
+
+            # TODO! parse list str when Quantity supports it
+            # s = repr([2, 3, 4] * u.kpc)[10:-1]
+            # num, unit = s.split("]")
+            # eval(num + "]") * u.Unit(unit)
+        meta = meta if meta else None  # empty meta -> None
+
+        # make cosmo instance
+        ba = cosmo_cls._init_signature.bind_partial(**params, meta=meta)
+        cosmo = cosmo_cls(*ba.args, **ba.kwargs)
+        return cosmo
+
+    def from_table(self, tbl):
+        """Instantiate a `~astropy.cosmology.Cosmology` from a |QTable|.
+
+        Parameters
+        ----------
+        tbl : `~astropy.table.QTable`
+
+        Returns
+        -------
+        `~astropy.cosmology.Cosmology` or list thereof
+
+        """
+        cosmos = [None] * len(tbl)
+        for i in range(len(tbl)):
+            # get metadata from table
+            name = tbl["name"][i] if "name" in tbl.colnames else None
+            meta = tbl.meta.get(name, {}) if name is not None else {}
```

Comment:
I think you'd need to add something like the following:
```
if name is masked:
    name = None
```

Reply:
Fantastic! That works great for csv.
Now testing the rest.

Reply:
Ahh. spoke a little too soon.

Reply:
Got it!
Thanks a bunch @mhvk !

Reply:
Happy it worked!